cuda/solver/gmres_kernels.cu

```diff
@@ -48,63 +57,687 @@ namespace cuda {
+        zero<ValueType>()) {
+        givens_cos[iter * stride_cos + col_idx] = zero<ValueType>();
+        givens_sin[iter * stride_sin + col_idx] = one<ValueType>();
+    } else {
+        auto hypotenuse =
+            sqrt(hessenberg_iter[iter * stride_hessenberg + col_idx] *
+                     hessenberg_iter[iter * stride_hessenberg + col_idx] +
+                 hessenberg_iter[(iter + 1) * stride_hessenberg + col_idx] *
+                     hessenberg_iter[(iter + 1) * stride_hessenberg + col_idx]);
+        givens_cos[iter * stride_cos + col_idx] =
+            abs(hessenberg_iter[iter * stride_hessenberg + col_idx]) /
+            hypotenuse;
+        givens_sin[iter * stride_sin + col_idx] =
+            givens_cos[iter * stride_cos + col_idx] *
+            hessenberg_iter[(iter + 1) * stride_hessenberg + col_idx] /
+            hessenberg_iter[iter * stride_hessenberg + col_idx];
+    }
+}
+
+
+template <typename ValueType>
+__device__ void calculate_residual_norm_kernel(
+    const size_type col_idx, const size_type num_cols,
+    const ValueType *givens_sin, size_type stride_sin,
+    const ValueType *givens_cos, size_type stride_cos, ValueType *residual_norm,
+    ValueType *residual_norm_collection,
+    size_type stride_residual_norm_collection, const ValueType *b_norm,
+    const size_type iter)
+{
+    residual_norm_collection[(iter + 1) * stride_residual_norm_collection +
+                             col_idx] =
+        -givens_sin[iter * stride_sin + col_idx] *
+        residual_norm_collection[iter * stride_residual_norm_collection +
+                                 col_idx];
+    residual_norm_collection[iter * stride_residual_norm_collection + col_idx] =
+        givens_cos[iter * stride_cos + col_idx] *
+        residual_norm_collection[iter * stride_residual_norm_collection +
+                                 col_idx];
+    residual_norm[col_idx] =
+        abs(residual_norm_collection[(iter + 1) *
+                                         stride_residual_norm_collection +
+                                     col_idx]) /
+        b_norm[col_idx];
+}
+
+
+// Must be called with at least `num_cols` threads in total.
+template <size_type block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void givens_rotation_kernel(
+    const size_type num_rows, const size_type num_cols,
+    ValueType *__restrict__ hessenberg_iter, size_type stride_hessenberg,
+    ValueType *__restrict__ givens_sin, size_type stride_sin,
+    ValueType *__restrict__ givens_cos, size_type stride_cos,
+    ValueType *__restrict__ residual_norm,
+    ValueType *__restrict__ residual_norm_collection,
+    size_type stride_residual_norm_collection,
+    const ValueType *__restrict__ b_norm, const size_type iter,
+    const stopping_status *__restrict__ stop_status)
+{
+    const auto col_idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if (col_idx >= num_cols || stop_status[col_idx].has_stopped()) return;
```

Comment:
Style: use `{}`. This also highlights the `return` more.

Reply:
Good catch, thanks!